File: scripts/a2a-discovery.py
#!/usr/bin/env python3
"""
A2A Agent Discovery Tool
Uses python-a2a SDK to discover agents via DiscoveryClient and AgentRegistry
Also includes domain probing for .well-known/agent-card.json
"""

import asyncio
import httpx
import json
import os
from datetime import datetime
from typing import List, Dict, Any, Optional

try:
    from python_a2a import A2AClient
    from python_a2a.discovery import DiscoveryClient, AgentRegistry
    HAS_A2A_SDK = True
except ImportError:
    HAS_A2A_SDK = False
    print("Warning: python-a2a SDK not fully available, using HTTP-only discovery")

KNOWN_REGISTRIES = [
    "https://registry.a2aprotocol.ai",
    "https://a2a-registry.truefoundry.com",
    "https://registry.artinet.io",
]

KNOWN_AGENT_DOMAINS = [
    "api.snack.money",
    "x402.arvos.xyz",
    "api.barvis.io",
    "api.jiren.ai",
    "api.dexter.cash",
    "api.canza.app",
    "x402-secure-api.t54.ai",
    "x402.lucyos.ai",
    "ainalyst-api.xyz",
    "pay.lnpay.ai",
    "mesh.heurist.xyz",
    "acp-x402.virtuals.io",
    "wurkapi.fun",
    "agents.memeputer.com",
    "firecrawl.dev",
    "otaku.so",
    "www.reap.deals",
    "www.qrbase.xyz",
]

SUBDOMAIN_PATTERNS = [
    "agent",
    "ai",
    "api",
    "a2a",
    "x402",
    "bot",
    "agents",
    "mcp",
]

TOP_DOMAINS_TO_PROBE = [
    "openai.com",
    "anthropic.com",
    "cohere.com",
    "google.com",
    "microsoft.com",
    "aws.amazon.com",
    "huggingface.co",
    "langchain.com",
    "llamaindex.ai",
    "crewai.com",
    "autogen.ai",
]

async def fetch_agent_card(client: httpx.AsyncClient, domain: str) -> Optional[Dict[str, Any]]:
    """Fetch agent card from a domain's .well-known path"""
    urls_to_try = [
        f"https://{domain}/.well-known/agent-card.json",
        f"https://{domain}/.well-known/agent.json",
        f"https://www.{domain}/.well-known/agent-card.json",
    ]
    
    for url in urls_to_try:
        try:
            response = await client.get(url, timeout=10.0, follow_redirects=True)
            if response.status_code == 200:
                try:
                    data = response.json()
                    print(f"✅ Found agent card at {url}")
                    return {"url": url, "domain": domain, "card": data}
                except json.JSONDecodeError:
                    pass
        except Exception as e:
            pass
    return None

async def probe_for_402(client: httpx.AsyncClient, domain: str) -> Optional[Dict[str, Any]]:
    """Probe a domain for x402 payment endpoints"""
    endpoints_to_try = [
        f"https://{domain}/",
        f"https://{domain}/api",
        f"https://{domain}/ping",
        f"https://{domain}/x402",
    ]
    
    for url in endpoints_to_try:
        try:
            response = await client.post(url, timeout=10.0, follow_redirects=True)
            if response.status_code == 402:
                try:
                    data = response.json()
                    if "x402Version" in data or "accepts" in data:
                        print(f"💰 Found x402 endpoint at {url}")
                        return {"url": url, "domain": domain, "x402_data": data}
                except json.JSONDecodeError:
                    pass
        except Exception as e:
            pass
    return None

async def query_registry(client: httpx.AsyncClient, registry_url: str) -> List[Dict[str, Any]]:
    """Query an A2A registry for agents"""
    agents = []
    try:
        response = await client.get(f"{registry_url}/agents", timeout=15.0)
        if response.status_code == 200:
            data = response.json()
            if isinstance(data, list):
                agents = data
            elif isinstance(data, dict) and "agents" in data:
                agents = data["agents"]
            print(f"📋 Found {len(agents)} agents in registry {registry_url}")
    except Exception as e:
        print(f"⚠️ Could not query registry {registry_url}: {e}")
    return agents

async def discover_agents_sdk() -> List[Dict[str, Any]]:
    """Use python-a2a SDK for discovery"""
    discovered = []
    if not HAS_A2A_SDK:
        return discovered
    
    try:
        for registry_url in KNOWN_REGISTRIES:
            try:
                registry = AgentRegistry(registry_url)
                agents = list(registry.get_all_agents())
                for agent in agents:
                    discovered.append({
                        "source": "a2a_sdk",
                        "registry": registry_url,
                        "agent": agent
                    })
                print(f"🔍 SDK discovered {len(agents)} agents from {registry_url}")
            except Exception as e:
                print(f"⚠️ SDK registry error {registry_url}: {e}")
    except Exception as e:
        print(f"⚠️ SDK discovery failed: {e}")
    
    return discovered

async def discover_via_subdomain_probing(client: httpx.AsyncClient) -> List[Dict[str, Any]]:
    """Probe subdomains of major domains for agent cards"""
    discovered = []
    
    for base_domain in TOP_DOMAINS_TO_PROBE:
        for pattern in SUBDOMAIN_PATTERNS:
            subdomain = f"{pattern}.{base_domain}"
            card = await fetch_agent_card(client, subdomain)
            if card:
                discovered.append({
                    "source": "subdomain_probe",
                    "type": "agent_card",
                    **card
                })
            
            x402 = await probe_for_402(client, subdomain)
            if x402:
                discovered.append({
                    "source": "subdomain_probe",
                    "type": "x402_endpoint",
                    **x402
                })
    
    return discovered

async def main():
    print("=" * 60)
    print("A2A Agent Discovery Tool")
    print(f"Started: {datetime.now().isoformat()}")
    print("=" * 60)
    
    all_discovered = []
    
    async with httpx.AsyncClient(
        headers={"User-Agent": "CoinRailz-AgentDiscovery/1.0"},
        verify=True
    ) as client:
        
        print("\n📡 Phase 1: Probing known agent domains...")
        tasks = [fetch_agent_card(client, domain) for domain in KNOWN_AGENT_DOMAINS]
        results = await asyncio.gather(*tasks)
        for result in results:
            if result:
                all_discovered.append({
                    "source": "known_domain",
                    "type": "agent_card",
                    **result
                })
        
        print(f"\n📡 Phase 2: Probing for x402 endpoints...")
        tasks = [probe_for_402(client, domain) for domain in KNOWN_AGENT_DOMAINS]
        results = await asyncio.gather(*tasks)
        for result in results:
            if result:
                all_discovered.append({
                    "source": "known_domain",
                    "type": "x402_endpoint",
                    **result
                })
        
        print(f"\n📡 Phase 3: Querying known registries...")
        for registry in KNOWN_REGISTRIES:
            agents = await query_registry(client, registry)
            for agent in agents:
                all_discovered.append({
                    "source": "registry",
                    "registry": registry,
                    "agent": agent
                })
        
        print(f"\n📡 Phase 4: Using A2A SDK discovery...")
        sdk_agents = await discover_agents_sdk()
        all_discovered.extend(sdk_agents)
        
        print(f"\n📡 Phase 5: Subdomain enumeration...")
        subdomain_agents = await discover_via_subdomain_probing(client)
        all_discovered.extend(subdomain_agents)
    
    print("\n" + "=" * 60)
    print(f"Discovery Complete: {len(all_discovered)} agents/endpoints found")
    print("=" * 60)
    
    output_file = "discovered_agents.json"
    with open(output_file, "w") as f:
        json.dump({
            "timestamp": datetime.now().isoformat(),
            "total_found": len(all_discovered),
            "agents": all_discovered
        }, f, indent=2, default=str)
    
    print(f"\n💾 Results saved to {output_file}")
    
    by_source = {}
    for agent in all_discovered:
        source = agent.get("source", "unknown")
        by_source[source] = by_source.get(source, 0) + 1
    
    print("\n📊 Summary by source:")
    for source, count in by_source.items():
        print(f"  - {source}: {count}")
    
    return all_discovered

if __name__ == "__main__":
    asyncio.run(main())

File: scripts/a2a-discovery.ts
#!/usr/bin/env npx tsx
/**
 * A2A Agent Discovery Tool (Node.js)
 * Uses @a2a-js/sdk for agent discovery
 * Includes domain probing, x402 detection, and registry queries
 */

import { A2AClient } from '@a2a-js/sdk';
import fs from 'fs';

interface DiscoveredAgent {
  source: string;
  type?: string;
  domain?: string;
  url?: string;
  card?: any;
  x402Data?: any;
  agent?: any;
  registry?: string;
}

const KNOWN_AGENT_DOMAINS = [
  'api.snack.money',
  'x402.arvos.xyz',
  'api.barvis.io',
  'api.jiren.ai',
  'api.dexter.cash',
  'api.canza.app',
  'x402-secure-api.t54.ai',
  'x402.lucyos.ai',
  'ainalyst-api.xyz',
  'pay.lnpay.ai',
  'mesh.heurist.xyz',
  'acp-x402.virtuals.io',
  'wurkapi.fun',
  'agents.memeputer.com',
  'firecrawl.dev',
  'otaku.so',
  'www.reap.deals',
  'www.qrbase.xyz',
];

const x402SCAN_TOP_SERVERS = [
  { domain: 'pay.lnpay.ai', volume: 7450, txns: 613800 },
  { domain: 'api.barvis.io', volume: 24190, txns: 401500 },
  { domain: 'x402.arvos.xyz', volume: 237660, txns: 382020 },
  { domain: 'x402-secure-api.t54.ai', volume: 52, txns: 368340 },
  { domain: 'api.jiren.ai', volume: 3480, txns: 348220 },
  { domain: 'api.canza.app', volume: 16350, txns: 293420 },
  { domain: 'api.dexter.cash', volume: 33940, txns: 198080 },
];

async function fetchWithTimeout(url: string, options: RequestInit = {}, timeout = 10000): Promise<Response> {
  const controller = new AbortController();
  const id = setTimeout(() => controller.abort(), timeout);
  
  try {
    const response = await fetch(url, {
      ...options,
      signal: controller.signal,
      headers: {
        'User-Agent': 'CoinRailz-AgentDiscovery/1.0',
        ...options.headers,
      },
    });
    clearTimeout(id);
    return response;
  } catch (error) {
    clearTimeout(id);
    throw error;
  }
}

async function fetchAgentCard(domain: string): Promise<DiscoveredAgent | null> {
  const urlsToTry = [
    `https://${domain}/.well-known/agent-card.json`,
    `https://${domain}/.well-known/agent.json`,
  ];
  
  for (const url of urlsToTry) {
    try {
      const response = await fetchWithTimeout(url);
      if (response.ok) {
        const data = await response.json();
        console.log(`✅ Found agent card at ${url}`);
        return { source: 'domain_probe', type: 'agent_card', domain, url, card: data };
      }
    } catch (e) {
      // Ignore errors
    }
  }
  return null;
}

async function probeFor402(domain: string): Promise<DiscoveredAgent | null> {
  const endpointsToTry = [
    `https://${domain}/`,
    `https://${domain}/api`,
    `https://${domain}/ping`,
  ];
  
  for (const url of endpointsToTry) {
    try {
      const response = await fetchWithTimeout(url, { method: 'POST' });
      if (response.status === 402) {
        const data = await response.json();
        if (data.x402Version || data.accepts) {
          console.log(`💰 Found x402 endpoint at ${url}`);
          return { source: 'domain_probe', type: 'x402_endpoint', domain, url, x402Data: data };
        }
      }
    } catch (e) {
      // Ignore errors
    }
  }
  return null;
}

async function discoverViaA2ASDK(): Promise<DiscoveredAgent[]> {
  const discovered: DiscoveredAgent[] = [];
  
  for (const serverInfo of x402SCAN_TOP_SERVERS) {
    try {
      const client = new A2AClient(`https://${serverInfo.domain}`);
      
      try {
        const agentCard = await client.agentCard();
        if (agentCard) {
          console.log(`🤖 A2A SDK found agent at ${serverInfo.domain}`);
          discovered.push({
            source: 'a2a_sdk',
            type: 'a2a_agent',
            domain: serverInfo.domain,
            agent: agentCard,
          });
        }
      } catch (e) {
        // Agent card fetch failed, try basic probe
      }
    } catch (e) {
      console.log(`⚠️ Could not connect to ${serverInfo.domain}`);
    }
  }
  
  return discovered;
}

async function scrapex402ScanServers(): Promise<DiscoveredAgent[]> {
  const discovered: DiscoveredAgent[] = [];
  
  try {
    const response = await fetchWithTimeout('https://www.x402scan.com/api/servers', {
      headers: { 'Accept': 'application/json' }
    });
    
    if (response.ok) {
      const data = await response.json();
      if (Array.isArray(data)) {
        for (const server of data) {
          discovered.push({
            source: 'x402scan_api',
            type: 'x402_server',
            domain: server.domain || server.host,
            agent: server,
          });
        }
        console.log(`📊 x402scan API returned ${data.length} servers`);
      }
    }
  } catch (e) {
    console.log('⚠️ x402scan API not accessible, using cached data');
    for (const server of x402SCAN_TOP_SERVERS) {
      discovered.push({
        source: 'x402scan_cache',
        type: 'x402_server',
        domain: server.domain,
        agent: server,
      });
    }
  }
  
  return discovered;
}

async function main() {
  console.log('='.repeat(60));
  console.log('A2A Agent Discovery Tool (Node.js)');
  console.log(`Started: ${new Date().toISOString()}`);
  console.log('='.repeat(60));
  
  const allDiscovered: DiscoveredAgent[] = [];
  
  console.log('\n📡 Phase 1: Probing known agent domains for agent cards...');
  const cardResults = await Promise.allSettled(
    KNOWN_AGENT_DOMAINS.map(domain => fetchAgentCard(domain))
  );
  for (const result of cardResults) {
    if (result.status === 'fulfilled' && result.value) {
      allDiscovered.push(result.value);
    }
  }
  
  console.log('\n📡 Phase 2: Probing for x402 endpoints...');
  const x402Results = await Promise.allSettled(
    KNOWN_AGENT_DOMAINS.map(domain => probeFor402(domain))
  );
  for (const result of x402Results) {
    if (result.status === 'fulfilled' && result.value) {
      allDiscovered.push(result.value);
    }
  }
  
  console.log('\n📡 Phase 3: Using A2A SDK for discovery...');
  const sdkAgents = await discoverViaA2ASDK();
  allDiscovered.push(...sdkAgents);
  
  console.log('\n📡 Phase 4: Scraping x402scan for active servers...');
  const x402scanAgents = await scrapex402ScanServers();
  allDiscovered.push(...x402scanAgents);
  
  console.log('\n' + '='.repeat(60));
  console.log(`Discovery Complete: ${allDiscovered.length} agents/endpoints found`);
  console.log('='.repeat(60));
  
  const output = {
    timestamp: new Date().toISOString(),
    totalFound: allDiscovered.length,
    agents: allDiscovered,
  };
  
  const outputFile = 'discovered_agents_node.json';
  fs.writeFileSync(outputFile, JSON.stringify(output, null, 2));
  console.log(`\n💾 Results saved to ${outputFile}`);
  
  const bySource: Record<string, number> = {};
  for (const agent of allDiscovered) {
    bySource[agent.source] = (bySource[agent.source] || 0) + 1;
  }
  
  console.log('\n📊 Summary by source:');
  for (const [source, count] of Object.entries(bySource)) {
    console.log(`  - ${source}: ${count}`);
  }
  
  return allDiscovered;
}

main().catch(console.error);
